Android: crash on start

```
FATAL EXCEPTION: main
Process: io.atticusc.atmosweather, PID: 24104
java.lang.RuntimeException: Unable to start activity ComponentInfo{io.atticusc.atmosweather/io.atticusc.atmosweather.MainActivity}: java.lang.SecurityException: Caller io.atticusc.atmosweather needs to hold android.permission.SCHEDULE_EXACT_ALARM or android.permission.USE_EXACT_ALARM to set exact alarms.
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3783)
	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3923)
	at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:103)
	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:139)
	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:96)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2444)
	at android.os.Handler.dispatchMessage(Handler.java:106)
	at android.os.Looper.loopOnce(Looper.java:205)
	at android.os.Looper.loop(Looper.java:294)
	at android.app.ActivityThread.main(ActivityThread.java:8223)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:552)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:977)
Caused by: java.lang.SecurityException: Caller io.atticusc.atmosweather needs to hold android.permission.SCHEDULE_EXACT_ALARM or android.permission.USE_EXACT_ALARM to set exact alarms.
	at android.os.Parcel.createExceptionOrNull(Parcel.java:3057)
	at android.os.Parcel.createException(Parcel.java:3041)
	at android.os.Parcel.readException(Parcel.java:3024)
	at android.os.Parcel.readException(Parcel.java:2966)
	at android.app.IAlarmManager$Stub$Proxy.set(IAlarmManager.java:311)
	at android.app.AlarmManager.setImpl(AlarmManager.java:1084)
	at android.app.AlarmManager.setImpl(AlarmManager.java:1044)
	at android.app.AlarmManager.setExact(AlarmManager.java:808)
	at io.atticusc.atmosweather.MainActivity.startBackgroundTask(MainActivity.java:86)
	at io.atticusc.atmosweather.MainActivity.onCreate(MainActivity.java:51)
	at android.app.Activity.performCreate(Activity.java:8595)
	at android.app.Activity.performCreate(Activity.java:8573)
	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1457)
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3765)
	... 12 more
Caused by: android.os.RemoteException: Remote stack trace:
	at com.android.server.alarm.AlarmManagerService$5.set(AlarmManagerService.java:2981)
	at android.app.IAlarmManager$Stub.onTransact(IAlarmManager.java:181)
	at android.os.Binder.execTransactInternal(Binder.java:1339)
	at android.os.Binder.execTransact(Binder.java:1275)
```

**Version Info**
 - Platform: Android REL 14 aarch64
 - Atmos Version: 2.0.1

--- a/android/app/src/main/AndroidManifest.xml
+++ b/android/app/src/main/AndroidManifest.xml
@@ -10,6 +10,8 @@
     <uses-permission android:name="android.permission.ACCESS_FINE_LOCATION" />
     <uses-permission android:name="android.permission.ACCESS_BACKGROUND_LOCATION" />
     <uses-permission android:name="android.permission.INTERNET" />
+    <uses-permission android:name="android.permission.POST_NOTIFICATIONS"/>
+
 
     <uses-feature
         android:name="android.hardware.location.gps"

--- a/android/app/src/main/java/io/atticusc/atmosweather/BackgroundService.java
+++ b/android/app/src/main/java/io/atticusc/atmosweather/BackgroundService.java
@@ -84,11 +84,25 @@ public class BackgroundService extends BroadcastReceiver {
             @SuppressLint("UnspecifiedImmutableFlag") PendingIntent pentent = PendingIntent.getBroadcast(context, 1, intentA, PendingIntent.FLAG_IMMUTABLE);
             Calendar calendar = Calendar.getInstance();
 
+            boolean canScheduleAlarms = true;
+            if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.S) {
+                canScheduleAlarms = alarmManager.canScheduleExactAlarms();
+            }
+
             // Update more often on WiFi
-            if (checkNetworkStatus(context)) {
-                alarmManager.setExact(AlarmManager.RTC_WAKEUP, calendar.getTimeInMillis() + 10000, pentent);
-            } else {
-                alarmManager.setExact(AlarmManager.RTC_WAKEUP, calendar.getTimeInMillis() + 25000, pentent);
+            if (canScheduleAlarms){
+                if (checkNetworkStatus(context)) {
+                    alarmManager.setExact(AlarmManager.RTC_WAKEUP, calendar.getTimeInMillis() + 10000, pentent);
+                } else {
+                    alarmManager.setExact(AlarmManager.RTC_WAKEUP, calendar.getTimeInMillis() + 25000, pentent);
+                }
+            }
+            else {
+                if (checkNetworkStatus(context)) {
+                    alarmManager.set(AlarmManager.RTC_WAKEUP, calendar.getTimeInMillis() + 10000, pentent);
+                } else {
+                    alarmManager.set(AlarmManager.RTC_WAKEUP, calendar.getTimeInMillis() + 25000, pentent);
+                }
             }
 
             JSONObject jObj = new JSONObject(weatherLocations.getString("settings", ""));

--- a/android/app/src/main/java/io/atticusc/atmosweather/MainActivity.java
+++ b/android/app/src/main/java/io/atticusc/atmosweather/MainActivity.java
@@ -11,6 +11,7 @@ import android.content.pm.PackageManager;
 import android.net.Uri;
 import android.os.Build;
 import android.os.Bundle;
+import android.provider.Settings;
 
 import androidx.core.app.ActivityCompat;
 
@@ -48,13 +49,18 @@ public class MainActivity extends BridgeActivity {
         // Cancel any duplicate alarms
         alarmManager.cancel(pendingIntent);
 
-        startBackgroundTask(pendingIntent, 5);
-
-        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.SCHEDULE_EXACT_ALARM) != PackageManager.PERMISSION_GRANTED) {
-            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
-                ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.SCHEDULE_EXACT_ALARM}, 1);
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
+            if (!alarmManager.canScheduleExactAlarms()){
+                Intent alarmIntent = new Intent(Settings.ACTION_REQUEST_SCHEDULE_EXACT_ALARM);
+                if (intent.resolveActivity(getPackageManager()) != null) {
+                    this.startActivity(alarmIntent);
+                }
             }
         }
+
+        startBackgroundTask(pendingIntent, 5);
+
+
 
         if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_BACKGROUND_LOCATION) != PackageManager.PERMISSION_GRANTED) {
             boolean locationInBackground = getLocationInBackgroundEnabled();

--- a/android/app/src/main/java/io/atticusc/atmosweather/MainActivity.java
+++ b/android/app/src/main/java/io/atticusc/atmosweather/MainActivity.java
@@ -50,6 +50,12 @@ public class MainActivity extends BridgeActivity {
 
         startBackgroundTask(pendingIntent, 5);
 
+        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.SCHEDULE_EXACT_ALARM) != PackageManager.PERMISSION_GRANTED) {
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
+                ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.SCHEDULE_EXACT_ALARM}, 1);
+            }
+        }
+
         if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_BACKGROUND_LOCATION) != PackageManager.PERMISSION_GRANTED) {
             boolean locationInBackground = getLocationInBackgroundEnabled();
 

--- a/android/app/src/main/java/io/atticusc/atmosweather/MainActivity.java
+++ b/android/app/src/main/java/io/atticusc/atmosweather/MainActivity.java
@@ -67,7 +67,12 @@ public class MainActivity extends BridgeActivity {
 
             if (locationInBackground) {
                 if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
-                    ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.ACCESS_BACKGROUND_LOCATION}, 1);
+                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
+                        ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.ACCESS_BACKGROUND_LOCATION, Manifest.permission.POST_NOTIFICATIONS}, 1);
+                    }
+                    else {
+                        ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.ACCESS_BACKGROUND_LOCATION}, 1);
+                    }
                 }
             }
         }

--- a/android/app/src/main/java/io/atticusc/atmosweather/MainActivity.java
+++ b/android/app/src/main/java/io/atticusc/atmosweather/MainActivity.java
@@ -95,7 +95,16 @@ public class MainActivity extends BridgeActivity {
         AlarmManager alarmManager = (AlarmManager) getSystemService(ALARM_SERVICE);
 
         Calendar calendar = Calendar.getInstance();
-        alarmManager.setExact(AlarmManager.RTC_WAKEUP, calendar.getTimeInMillis() + (delaySeconds * 1000L), pendingIntent);
+        boolean canScheduleAlarms = true;
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
+            canScheduleAlarms = alarmManager.canScheduleExactAlarms();
+        }
+        if (canScheduleAlarms){
+            alarmManager.setExact(AlarmManager.RTC_WAKEUP, calendar.getTimeInMillis() + (delaySeconds * 1000L), pendingIntent);
+        }
+        else {
+            alarmManager.set(AlarmManager.RTC_WAKEUP, calendar.getTimeInMillis() + (delaySeconds * 1000L), pendingIntent);
+        }
     }
 
     private boolean isFirstRun() {

